### PR TITLE
Simple service proxy & Key-Value cache module

### PR DESCRIPTION
**Pull request by gwicke**
Bug: https://phabricator.wikimedia.org/T95551

### CODE REVIEW

File: mods/key_rev_value.js [hunk 1]
```diff
@@ -50,7 +50,7 @@ KRVBucket.prototype.makeSchema = function (opts) {
             'content-sha256': 'blob',
             // redirect
             'content-location': 'string',
-            tags: 'set<string>'
+            tags: 'set<string>',
```
Comment (d00rman) [inline mods/key_rev_value.js]: trailing comma?

File: mods/simple_service.js [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+function SimpleService(options) {
+    options = options || {};
+    this.spec = {
+        paths: options.paths
+    };
+
+    this.exports = this.processSpec(this.spec);
+}
+
+var headerWhitelist = ['accept', 'accept-language', 'accept-language'];
```
Comment (d00rman) [inline mods/simple_service.js]: two `accept-language`

File: mods/simple_service.js [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                        })
+                        .then(function(storeRes) {
+                            res.headers.etag = storeRes.headers.etag;
+                            return res;
+                        });
+                    });
+                }
+
+                if (conf.storage) {
+                    if (conf.storage['no-cache_refresh']
```
Comment (d00rman) [inline mods/simple_service.js]: `no-cache_refresh` is a rather awkward syntax to me. How about `nocache_refresh` ?

File: package.json [hunk 4]
```diff
@@ -42,8 +42,8 @@
     "preq": "~0.3.6",
     "request": "^2.44.0",
     "restbase-mod-table-cassandra": "^0.5.3",
-    "service-runner": "^0.1.6",
-    "swagger-router": "^0.0.4",
+    "service-runner": "^0.1.7",
```
Comment (d00rman) [inline package.json]: we're at `0.1.8` now ;)

File: test/features/simple_service.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        })
+        .then(function (res) {
+            // Since this is a dynamic page which should render the same each
+            // time, the tid should not change.
+            tid2 = res.headers.etag;
+            assert.notDeepEqual(tid2, tid1);
+            assert.notDeepEqual(tid2, undefined);
+            hasTextContentType(res);
+        });
+    });
```
Comment (d00rman) [inline test/features/simple_service.js]: IMO It'd be good to add a second test retrieving the page from cache and checking that no remote requests have been made.

**Comment by d00rman** [comment]
LGTM, small nitpicks in-lined.

In general, I am not happy about the fact that we will still have to add a new module (or path in the config) for each proxied service. It would be easier, IMO, if there could be a way to specify which URL to call directly in the public spec, instead of relying on the cache module to provide that. So, in `specs/test.yaml`:

```yaml
paths:
  /{module:service}/test/{title}{/rev}:
    get:
      x-backend-request:
        uri: /{domain}/sys/testservice/test/{title}{/revision}
        headers:
          x-uri: http://en.wikipedia.org/wiki/{title}
```
Then, in the module def in `config.yaml`, instead of having the path `/test/{key}`, we could have `/{component}/{key}`. `x-uri` could be picked up by the cache module.

That would allow us to have one path per cache type in the config. That would also warrant on-the-fly bucket creation, though.

**Comment by gwicke** [comment]
@d00rman, so far we have kept all local configs like backend service ips etc out of the public spec. The only rewriting that happens there is the one to the internal service end point. I think this makes sense to keep it that way. Advantages I see are:

- No need to puppetize the public spec
- Can change the public spec without breaking code that relies on the service internally; can expose the internal service in multiple ways / with different frontend entry points
- Can route a public end point to a different module at any time, for example if the service develops special needs that don't fit into the regular cache implementation

File: package.json (hunk 4 above)
Comment (gwicke) [inline package.json]: Okay, that's really a separate issue, but I bumped it in this PR nevertheless.

File: mods/simple_service.js (hunk 3 above)
Comment (gwicke) [inline mods/simple_service.js]: The `no-cache` part reflects the header. I agree that it doesn't look pretty, but at least it's consistent with our policy of using _ as a delimiter between components.

File: mods/key_rev_value.js (hunk 1 above)
Comment (gwicke) [inline mods/key_rev_value.js]: Yup, slowly moving towards https://github.com/felixge/node-style-guide

**Comment by gwicke** [comment]
Rebased onto master.

**Comment by gwicke** [comment]
Removed the WIP from the first patch in the series.

File: mods/simple_service.js (hunk 3 above)
Comment (d00rman) [inline mods/simple_service.js]: I understand the rationale, but I still feel it's awkward and error-prone.

**Pull request merged**